File: src/content/docs/topics/cs/coding-problems/1d-dynamic-programming/416-partition-equal-subset-sum.py
def can_partition(nums):
    pass  # TODO: implement

def _run_tests():
    assert can_partition([1, 5, 11, 5]) == True
    assert can_partition([1, 2, 3, 5]) == False
    assert can_partition([1]) == False
    assert can_partition([2, 2]) == True
    assert can_partition([1, 2, 5]) == False
    assert can_partition([3, 3, 3, 4, 5]) == True
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    can_partition(list(range(1, 101)))
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf can_partition(1..100): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/1d-dynamic-programming/416-partition-equal-subset-sum.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func canPartition(_ nums: [Int]) -> Bool {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:549a768aeb46e811fc6e29139a3bb4669bdd81b27bce75a1d58fd75ebc4a9f8f
    expectEqual(Solution().canPartition([1, 5, 11, 5]), true, "partitionable")
    expectEqual(Solution().canPartition([1, 2, 3, 5]), false, "not-partitionable")
    expectEqual(Solution().canPartition([2, 2, 1, 1]), true, "duplicates")
    expectEqual(Solution().canPartition([1]), false, "single-value")
    // EXCLUDED_VECTOR empty-array: [[]] | The published input contains at least one value.
    // EXCLUDED_VECTOR zero-value: [[0,1]] | Published values are positive.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
